sdk/sdkintegrations/initconn.go

```diff
@@ -5,37 +5,86 @@ import (
 	"net/http"
 	"net/url"
 
+	"go.uber.org/zap"
+
 	"go.autokitteh.dev/autokitteh/internal/kittehs"
 	"go.autokitteh.dev/autokitteh/sdk/sdktypes"
 )
 
-// FinalizeConnectionInit finalizes all integration-specific connection flows.
-// It encodes their resulting details and redirects the user to the final HTTP
-// handler ("post-init") that saves the data in the connection's scope, and
-// redirects the user to the last HTTP response, based on its origin
-// (local AK server / local VS Code extension / SaaS web UI).
-func FinalizeConnectionInit(w http.ResponseWriter, r *http.Request, iid sdktypes.IntegrationID, data []sdktypes.Var) {
-	vars, err := kittehs.EncodeURLData(data)
-	if err != nil {
-		http.Error(w, "Failed to encode URL vars", http.StatusInternalServerError)
-		return
+type ConnectionInit struct {
+	Writer       http.ResponseWriter
+	Request      *http.Request
+	Integration  sdktypes.Integration
+	ConnectionID string
+	Origin       string
+	logger       *zap.Logger
+}
+
+func NewConnectionInit(l *zap.Logger, w http.ResponseWriter, r *http.Request, i sdktypes.Integration) (ConnectionInit, *zap.Logger) {
+	cid := r.FormValue("cid")
+	origin := r.FormValue("origin")
+
+	l = l.With(
+		zap.String("urlPath", r.URL.Path),
+		zap.String("connectionID", cid),
+		zap.String("origin", origin),
+	)
+
+	return ConnectionInit{
+		Writer:       w,
+		Request:      r,
+		Integration:  i,
+		ConnectionID: cid,
+		Origin:       origin,
+		logger:       l,
+	}, l
+}
+
+// Abort is the same as [AbortWithStatus] with HTTP 400 (Bad Request).
+func (c ConnectionInit) Abort(err string) {
+	c.AbortWithStatus(http.StatusBadRequest, err)
+}
+
+// Abort aborts the connection initialization flow due to
+// a runtime error. It encodes the error status and message,
+// and redirects the user to the last HTTP response, based on its
+// origin (local AK server / local VS Code extension / SaaS web UI).
+func (c ConnectionInit) AbortWithStatus(status int, err string) {
+	switch c.Request.FormValue("origin") {
+	case "vscode":
+		u := "vscode://autokitteh.autokitteh?cid=%s&status=%d&error=%s"
+		u = fmt.Sprintf(u, c.ConnectionID, status, url.QueryEscape(err))
+		http.Redirect(c.Writer, c.Request, u, http.StatusFound)
+	case "web":
+		u := "/connections/%s/result?status=%d&error=%s"
+		u = fmt.Sprintf(u, c.ConnectionID, status, url.QueryEscape(err))
+		http.Redirect(c.Writer, c.Request, u, http.StatusFound)
+	default: // Local server ("cli", "dash", etc.)
+		path := c.Integration.ConnectionURL().Path + "/error.html"
```

Comment:
You need to put the right status code here, not just found. If there is an error, the right error code should be returned. 
Also - how do you make Ronen parse the error?

Reply:
The behavior that Ronen asked for was to always get HTTP 200, and parse the status from a JSON body - but you're right. I confirmed with Ronen that HTTP status + error message in the HTTP response's body is in fact better.

Reply:
No, we need to return proper status codes. I also told this to Ronen.

Reply:
Yes - already done, see line 59